File: lib/utils/sound-change.ts
/**
 * Sound Change Pipeline Engine
 *
 * Parses and applies phonological rules using standard linguistic notation:
 *   target → replacement / left_environment _ right_environment
 *
 * Examples:
 *   a → e / _#          (a becomes e word-finally)
 *   k → tʃ / _i         (k becomes tʃ before i)
 *   s → ∅ / V_V         (s is deleted between vowels)
 *   b → p / #_           (b becomes p word-initially)
 *
 * Special symbols:
 *   # = word boundary
 *   V = any vowel
 *   C = any consonant
 *   ∅ (or 0) = deletion
 *   _ = position of the target
 */

// Default phoneme classes — comprehensive IPA inventory
const DEFAULT_VOWELS = new Set([
  // Close
  "i", "y", "ɨ", "ʉ", "ɯ", "u",
  // Near-close
  "ɪ", "ʏ", "ʊ",
  // Close-mid
  "e", "ø", "ɘ", "ɵ", "ɤ", "o",
  // Mid
  "ə", "e̞", "ø̞", "ɤ̞", "o̞",
  // Open-mid
  "ɛ", "œ", "ɜ", "ɞ", "ʌ", "ɔ",
  // Near-open
  "æ", "ɐ",
  // Open
  "a", "ɶ", "ä", "ɑ", "ɒ",
  // Nasalized
  "ã", "ẽ", "ĩ", "õ", "ũ",
  // Long
  "aː", "eː", "iː", "oː", "uː", "ɛː", "ɔː", "æː", "ɑː",
  // Rhoticized
  "ɚ", "ɝ",
])

const DEFAULT_CONSONANTS = new Set([
  // Plosives
  "p", "b", "t", "d", "ʈ", "ɖ", "c", "ɟ", "k", "g", "ɡ", "q", "ɢ", "ʔ", "ʡ",
  // Nasals
  "m", "ɱ", "n", "ɳ", "ɲ", "ŋ", "ɴ",
  // Trills
  "ʙ", "r", "ʀ", "ʜ", "ʢ",
  // Taps/Flaps
  "ⱱ", "ɾ", "ɽ", "ɺ",
  // Fricatives
  "ɸ", "β", "f", "v", "θ", "ð", "s", "z", "ʃ", "ʒ", "ʂ", "ʐ",
  "ç", "ʝ", "x", "ɣ", "χ", "ʁ", "ħ", "ʕ", "h", "ɦ",
  "ɕ", "ʑ", "ʍ", "ɧ", "ɼ",
  // Lateral fricatives
  "ɬ", "ɮ",
  // Affricates
  "ts", "dz", "tʃ", "dʒ", "tɕ", "dʑ", "ʈʂ", "ɖʐ", "pf", "bv", "kx", "gɣ", "ɡɣ",
  // Approximants
  "ʋ", "ɹ", "ɻ", "j", "ɰ", "w", "ɥ",
  // Lateral approximants
  "l", "ɫ", "ɭ", "ʎ", "ʟ",
  // Implosives
  "ɓ", "ɗ", "ʄ", "ɠ", "ʛ",
])

export interface SoundChangeRule {
  id: string
  /** Raw text of the rule, e.g. "a → e / _#" */
  raw: string
  /** The target segment(s) to match */
  target: string
  /** The replacement (empty string for deletion) */
  replacement: string
  /** Left environment regex pattern */
  leftEnv: string
  /** Right environment regex pattern */
  rightEnv: string
  /** Whether the rule is enabled */
  enabled: boolean
  /** Optional note / description */
  note?: string
}

export interface SoundChangeResult {
  original: string
  changed: string
  rulesApplied: string[]
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
}

/**
 * Build a `(?:…|…)` alternation for a phoneme class, longest-first so multi-char
 * phonemes (e.g. "aː", "tʃ") match before their single-char prefixes.
 *
 * Cached per Set: these alternations are ~50 entries each and were previously
 * rebuilt on every rule application — a hot path during `batchApply`.
 */
const classAlternationCache = new WeakMap<Set<string>, string>()

function classAlternation(set: Set<string>): string {
  const cached = classAlternationCache.get(set)
  if (cached) return cached
  const alt = Array.from(set)
    .sort((a, b) => b.length - a.length)
    .map(escapeRegex)
    .join("|")
  const pattern = `(?:${alt})`
  classAlternationCache.set(set, pattern)
  return pattern
}

interface EnvPattern {
  /** Regex source for the environment's characters (no boundary anchor). */
  pattern: string
  /** Whether the environment requires a word boundary (#) at the word edge. */
  boundary: boolean
}

/**
 * Cache of user-class-name lists sorted longest-first, keyed by the Map
 * instance. Mirrors `classAlternationCache`: names are recomputed once per
 * distinct classes Map rather than on every `envToPattern` call.
 */
const sortedClassNamesCache = new WeakMap<Map<string, Set<string>>, string[]>()

function sortedClassNames(classes: Map<string, Set<string>>): string[] {
  const cached = sortedClassNamesCache.get(classes)
  if (cached) return cached
  const names = Array.from(classes.keys()).sort((a, b) => b.length - a.length)
  sortedClassNamesCache.set(classes, names)
  return names
}

/**
 * Convert an environment string to a regex source plus a boundary flag.
 * Handles: V (vowel), C (consonant), # (boundary), . (any), user-defined
 * named classes, literal characters.
 *
 * User class names are multi-char, so this can't be a naive per-char loop
 * for them: we walk the env string with a cursor and, at each position, try
 * every registered class name (longest first) as a literal prefix match
 * before falling back to the single-char built-in handling. This guarantees
 * a class named "KW" is consumed as one token rather than "K" + literal "W"
 * when both "K" and "KW" are defined.
 *
 * The boundary position (start vs end) is applied by the caller so the
 * result can be used as a zero-width lookbehind (left) or lookahead (right).
 */
function envToPattern(
  env: string,
  vowels: Set<string>,
  consonants: Set<string>,
  classes?: Map<string, Set<string>>
): EnvPattern {
  if (!env || env === "_") return { pattern: "", boundary: false }

  const names = classes && classes.size > 0 ? sortedClassNames(classes) : []

  let pattern = ""
  let boundary = false
  let i = 0
  outer: while (i < env.length) {
    for (const name of names) {
      if (env.startsWith(name, i)) {
        pattern += classAlternation(classes!.get(name)!)
        i += name.length
        continue outer
      }
    }

    const ch = env[i]
    if (ch === "#") {
      boundary = true
    } else if (ch === "V") {
      pattern += classAlternation(vowels)
    } else if (ch === "C") {
      pattern += classAlternation(consonants)
    } else if (ch === ".") {
      pattern += "."
    } else {
      pattern += escapeRegex(ch)
    }
    i += 1
  }
  return { pattern, boundary }
}

/**
 * Parse a sound change rule from text.
 *
 * Formats supported:
 *   target → replacement / left _ right
 *   target > replacement / left _ right
 *   target -> replacement / left _ right
 *   target → replacement   (no environment = unconditional)
 */
export function parseRule(text: string, id?: string): SoundChangeRule | null {
  const trimmed = text.trim()
  if (!trimmed || trimmed.startsWith("//") || trimmed.startsWith("#")) {
    return null // Comment or empty line
  }

  // Split on arrow. Accepts →, a single >, or one-or-more dashes/equals before
  // `>` so the natural ASCII forms `->`, `-->`, `==>` all work — plus the
  // en/em-dash variants (–> —>) macOS "smart dashes" can produce. The dash run
  // is matched as a unit so `-->` doesn't leave a stray `-` glued to the target.
  const arrowMatch = trimmed.match(/^(.+?)(?:→|[=–—-]+>|>)(.+)$/)
  if (!arrowMatch) return null

  const target = arrowMatch[1].trim()
  let rest = arrowMatch[2].trim()

  // Split on / for environment
  let replacement = ""
  let leftEnv = ""
  let rightEnv = ""

  const envSplit = rest.split("/")
  replacement = envSplit[0].trim()

  if (envSplit.length > 1) {
    const envPart = envSplit.slice(1).join("/").trim()
    const underscoreIdx = envPart.indexOf("_")
    if (underscoreIdx >= 0) {
      leftEnv = envPart.substring(0, underscoreIdx).trim()
      rightEnv = envPart.substring(underscoreIdx + 1).trim()
    }
  }

  // Handle deletion symbols
  if (replacement === "∅" || replacement === "0" || replacement === "Ø") {
    replacement = ""
  }

  return {
    id: id || crypto.randomUUID(),
    raw: trimmed,
    target,
    replacement,
    leftEnv,
    rightEnv,
    enabled: true,
  }
}

/**
 * Apply a single rule to a word.
 */
export function applyRule(
  word: string,
  rule: SoundChangeRule,
  vowels: Set<string> = DEFAULT_VOWELS,
  consonants: Set<string> = DEFAULT_CONSONANTS,
  classes?: Map<string, Set<string>>
): string {
  if (!rule.enabled || !rule.target) return word

  const left = envToPattern(rule.leftEnv, vowels, consonants, classes)
  const right = envToPattern(rule.rightEnv, vowels, consonants, classes)
  const targetClass = classes?.get(rule.target)
  const target = targetClass ? classAlternation(targetClass) : `(?:${escapeRegex(rule.target)})`

  // Environments are expressed as ZERO-WIDTH assertions — a lookbehind for the
  // left context and a lookahead for the right — so only the target segment is
  // consumed. This is what makes adjacent/overlapping contexts work: a rule like
  // `s → ∅ / V_V` over "asasa" must delete BOTH intervocalic consonants, but a
  // regex that *consumes* the shared vowel can only match every other target.
  // `#` anchors to the word start (left) or end (right).
  const lookbehind =
    left.boundary && !left.pattern ? "^"
    : left.boundary ? `(?<=^${left.pattern})`
    : left.pattern ? `(?<=${left.pattern})`
    : ""

  const lookahead =
    right.boundary && !right.pattern ? "$"
    : right.boundary ? `(?=${right.pattern}$)`
    : right.pattern ? `(?=${right.pattern})`
    : ""

  try {
    const regex = new RegExp(`${lookbehind}${target}${lookahead}`, "g")
    return word.replace(regex, rule.replacement)
  } catch {
    // Invalid regex — return unchanged
    return word
  }
}

/**
 * Apply a pipeline of rules to a single word.
 * Rules are applied sequentially — order matters!
 */
export function applyPipeline(
  word: string,
  rules: SoundChangeRule[],
  vowels?: Set<string>,
  consonants?: Set<string>,
  classes?: Map<string, Set<string>>
): SoundChangeResult {
  let current = word
  const rulesApplied: string[] = []

  for (const rule of rules) {
    if (!rule.enabled) continue
    const result = applyRule(current, rule, vowels, consonants, classes)
    if (result !== current) {
      rulesApplied.push(rule.raw)
      current = result
    }
  }

  return {
    original: word,
    changed: current,
    rulesApplied,
  }
}

/**
 * Parse multiple rules from a text block (one per line).
 * Lines starting with // or # are comments and are skipped.
 */
export function parseRules(text: string): SoundChangeRule[] {
  return text
    .split("\n")
    .map((line, i) => parseRule(line, `rule-${i}`))
    .filter((r): r is SoundChangeRule => r !== null)
}

/**
 * Batch-apply a pipeline to multiple words.
 */
export function batchApply(
  words: string[],
  rules: SoundChangeRule[],
  vowels?: Set<string>,
  consonants?: Set<string>,
  classes?: Map<string, Set<string>>
): SoundChangeResult[] {
  return words.map(word => applyPipeline(word, rules, vowels, consonants, classes))
}

/**
 * A parsed rule-set source: user-defined named sound classes plus the
 * ordered rules that reference them (or don't).
 */
export interface SoundChangeProgram {
  classes: Map<string, Set<string>>
  rules: SoundChangeRule[]
}

/** Identifier for a class name: a non-empty run of letters only. */
const CLASS_NAME_RE = /^[A-Za-z]+$/

/** Class names "V" and "C" are reserved — built-in vowel/consonant tokens take precedence. */
const RESERVED_CLASS_NAMES = new Set(["V", "C"])

/**
 * Parse a single `class NAME = m1 m2 m3` definition line.
 * Returns null for anything else, including malformed class lines, so
 * callers can fall through to `parseRule`.
 */
function parseClassLine(line: string): { name: string; members: Set<string> } | null {
  const trimmed = line.trim()
  if (!trimmed) return null

  const prefix = "class "
  if (!trimmed.startsWith(prefix)) return null
  const rest = trimmed.slice(prefix.length)

  const eq = rest.indexOf("=")
  if (eq < 0) return null

  const name = rest.slice(0, eq).trim()
  if (!name || !CLASS_NAME_RE.test(name) || RESERVED_CLASS_NAMES.has(name)) return null

  const members = rest
    .slice(eq + 1)
    .split(/\s+/)
    .map(m => m.trim())
    .filter(m => m.length > 0)
  if (members.length === 0) return null

  return { name, members: new Set(members) }
}

/**
 * Parse a rule-set source into user-defined named sound classes plus the
 * ordered rules that reference them (or don't). Lines of the form
 * `class NAME = m1 m2 m3` are collected into `classes` (last definition wins
 * on duplicate names); every other line is parsed with the existing
 * `parseRule`, exactly as `parseRules` already does, so class-free text
 * produces `rules` identical to `parseRules`.
 */
export function parseProgram(text: string): SoundChangeProgram {
  const classes = new Map<string, Set<string>>()
  const rules: SoundChangeRule[] = []

  text.split("\n").forEach((line, i) => {
    const classDef = parseClassLine(line)
    if (classDef) {
      classes.set(classDef.name, classDef.members)
      return
    }
    const rule = parseRule(line, `rule-${i}`)
    if (rule) rules.push(rule)
  })

  return { classes, rules }
}

File: linguistics-core/soundchange/soundchange.go
// Package soundchange is a phoneme-aware sound-change engine: it parses
// phonological rules (target → replacement / left _ right) and applies them to
// words.
//
// It is a Go port of lib/utils/sound-change.ts. The key design difference: the
// TS version uses regex with lookbehind/lookahead, but Go's stdlib regexp (RE2)
// has no lookarounds. So this engine uses a direct position scanner instead —
// which is faster (no per-word regex compilation), correct by construction for
// adjacent/overlapping environments (e.g. intervocalic `s → ∅ / V_V` deletes
// BOTH consonants in "asasa"), and runs identically on every platform via WASM
// (no Safari-version lookbehind caveat).
package soundchange

import (
	"regexp"
	"sort"
	"strings"
)

// arrowRe matches the rule arrow: →, a single >, or one-or-more dashes/equals
// before `>` (so `->`, `-->`, `==>` work) plus the en/em-dash variants (–> —>)
// macOS "smart dashes" can produce. The dash run is matched as a unit so `-->`
// does not leave a stray `-` glued to the target. Mirrors the TS engine regex.
var arrowRe = regexp.MustCompile(`→|[=–—-]+>|>`)

// Rule is a single phonological rewrite rule.
type Rule struct {
	Raw         string `json:"raw"`
	Target      string `json:"target"`
	Replacement string `json:"replacement"`
	LeftEnv     string `json:"leftEnv"`
	RightEnv    string `json:"rightEnv"`
	Enabled     bool   `json:"enabled"`
}

// Result is the outcome of applying a pipeline to one word.
type Result struct {
	Original     string   `json:"original"`
	Changed      string   `json:"changed"`
	RulesApplied []string `json:"rulesApplied"`
}

type tokenKind uint8

const (
	tokVowel tokenKind = iota
	tokConsonant
	tokBoundary // #
	tokAny      // .
	tokLiteral
	tokClass // user-defined named class (matches any one member)
)

type token struct {
	kind  tokenKind
	lit   []rune   // for tokLiteral
	class [][]rune // for tokClass: members, longest-first
}

// Engine holds the phoneme classes (vowels/consonants), stored longest-first so
// multi-rune phonemes (e.g. "aː", "tʃ") match before their prefixes.
type Engine struct {
	vowels     [][]rune
	consonants [][]rune
	// userClasses maps a class name to its members (longest-first).
	userClasses map[string][][]rune
	// userClassNames holds userClasses' keys sorted longest-first,
	// precomputed once at construction so tokenize never re-sorts this
	// invariant list on the hot path.
	userClassNames []string
}

// NewEngine builds an engine from explicit vowel/consonant inventories.
func NewEngine(vowels, consonants []string) *Engine {
	return &Engine{
		vowels:     toSortedRunes(vowels),
		consonants: toSortedRunes(consonants),
	}
}

// NewEngineWithClasses builds an engine from explicit vowel/consonant
// inventories plus user-defined named sound classes (as parsed into
// Program.Classes). Each class's members are stored longest-first, exactly
// like vowels/consonants, so multi-rune members match before their prefixes.
//
// An engine built with a nil/empty classes map behaves identically to
// NewEngine, since tokenize/ApplyRule only special-case names present in
// userClasses.
func NewEngineWithClasses(vowels, consonants []string, classes map[string][]string) *Engine {
	return NewEngine(vowels, consonants).withClasses(classes)
}

// withClasses registers user classes on an already-built engine and returns
// it. Members are stored longest-first, and the sorted class-name list is
// precomputed once (invariant for the engine's lifetime, so tokenize never
// re-sorts on the hot path). A nil/empty map is a no-op, leaving the engine
// class-free. Shared by NewEngineWithClasses and
// NewEngineOrDefaultWithClasses so the precompute logic lives in one place.
func (e *Engine) withClasses(classes map[string][]string) *Engine {
	if len(classes) == 0 {
		return e
	}
	e.userClasses = make(map[string][][]rune, len(classes))
	for name, members := range classes {
		e.userClasses[name] = toSortedRunes(members)
	}
	e.userClassNames = sortedClassNames(e.userClasses)
	return e
}

// DefaultEngine uses the comprehensive default IPA inventory.
func DefaultEngine() *Engine {
	return NewEngine(defaultVowels, defaultConsonants)
}

// NewEngineOrDefault builds an engine, falling back to the default inventory for
// any class left empty. This mirrors the TS engine, where an omitted
// vowels/consonants argument independently defaults that one class.
func NewEngineOrDefault(vowels, consonants []string) *Engine {
	v, c := defaultedInventories(vowels, consonants)
	return NewEngine(v, c)
}

// NewEngineOrDefaultWithClasses defaults empty vowel/consonant inventories to
// the built-in IPA sets (exactly like NewEngineOrDefault) AND registers the
// given user-defined classes. It shares the class-registration path with
// NewEngineWithClasses, so userClassNames is still precomputed once.
func NewEngineOrDefaultWithClasses(vowels, consonants []string, classes map[string][]string) *Engine {
	v, c := defaultedInventories(vowels, consonants)
	return NewEngine(v, c).withClasses(classes)
}

// defaultedInventories returns vowels/consonants with each empty class
// independently replaced by its built-in default set.
func defaultedInventories(vowels, consonants []string) (v, c []string) {
	v, c = vowels, consonants
	if len(v) == 0 {
		v = defaultVowels
	}
	if len(c) == 0 {
		c = defaultConsonants
	}
	return v, c
}

func toSortedRunes(set []string) [][]rune {
	out := make([][]rune, 0, len(set))
	for _, s := range set {
		out = append(out, []rune(s))
	}
	sort.SliceStable(out, func(i, j int) bool { return len(out[i]) > len(out[j]) })
	return out
}

// ParseRule parses one rule line. Returns ok=false for comments/blank/invalid.
func ParseRule(text string) (Rule, bool) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" || strings.HasPrefix(trimmed, "//") || strings.HasPrefix(trimmed, "#") {
		return Rule{}, false
	}

	// Split on the first arrow: →, ->, or >
	target, rest, ok := splitArrow(trimmed)
	if !ok {
		return Rule{}, false
	}
	target = strings.TrimSpace(target)

	replacement := strings.TrimSpace(rest)
	leftEnv, rightEnv := "", ""
	if idx := strings.Index(rest, "/"); idx >= 0 {
		replacement = strings.TrimSpace(rest[:idx])
		envPart := strings.TrimSpace(rest[idx+1:])
		if u := strings.Index(envPart, "_"); u >= 0 {
			leftEnv = strings.TrimSpace(envPart[:u])
			rightEnv = strings.TrimSpace(envPart[u+1:])
		}
	}

	switch replacement {
	case "∅", "0", "Ø":
		replacement = ""
	}

	return Rule{
		Raw:         trimmed,
		Target:      target,
		Replacement: replacement,
		LeftEnv:     leftEnv,
		RightEnv:    rightEnv,
		Enabled:     true,
	}, true
}

func splitArrow(s string) (left, right string, ok bool) {
	loc := arrowRe.FindStringIndex(s)
	if loc == nil {
		return "", "", false
	}
	return s[:loc[0]], s[loc[1]:], true
}

// ParseRules parses a newline-separated block, skipping comments/blanks.
func ParseRules(text string) []Rule {
	var rules []Rule
	for _, line := range strings.Split(text, "\n") {
		if r, ok := ParseRule(line); ok {
			rules = append(rules, r)
		}
	}
	return rules
}

// tokenize splits an environment string into matcher tokens. classNames is
// the class-name list already sorted longest-first (precomputed once on the
// Engine), and classes maps each name to its (longest-first) members; at each
// position we try to match the longest defined class name before falling
// back to the built-in V/C/#/. tokens or a literal rune. Nil classNames/
// classes make this behave exactly as it did before user classes existed.
func tokenize(env string, classNames []string, classes map[string][][]rune) []token {
	runes := []rune(env)
	var toks []token
	for i := 0; i < len(runes); {
		// classNames is sorted longest-first, so the first name that is a
		// rune-prefix here is the longest match. Two distinct class names
		// cannot both match at this position with equal rune length — the
		// substring of a given length at a given offset is unique — so map
		// iteration order (and thus tie-break stability) is not
		// load-bearing: tokenization is deterministic regardless.
		if name, n := matchLongestClassName(runes, i, classNames); n > 0 {
			toks = append(toks, token{kind: tokClass, class: classes[name]})
			i += n
			continue
		}
		switch runes[i] {
		case '#':
			toks = append(toks, token{kind: tokBoundary})
		case 'V':
			toks = append(toks, token{kind: tokVowel})
		case 'C':
			toks = append(toks, token{kind: tokConsonant})
		case '.':
			toks = append(toks, token{kind: tokAny})
		default:
			toks = append(toks, token{kind: tokLiteral, lit: []rune{runes[i]}})
		}
		i++
	}
	return toks
}

// sortedClassNames returns the class names sorted longest-first (by rune
// count), so a name that is a prefix of another (e.g. "K" vs "KW") never
// shadows the longer match.
func sortedClassNames(classes map[string][][]rune) []string {
	if len(classes) == 0 {
		return nil
	}
	names := make([]string, 0, len(classes))
	for name := range classes {
		names = append(names, name)
	}
	sort.SliceStable(names, func(i, j int) bool {
		return len([]rune(names[i])) > len([]rune(names[j]))
	})
	return names
}

// matchLongestClassName returns the first (longest, by construction of
// names) class name that is a rune-prefix of runes[p:], and its rune length.
// Returns ("", 0) if none matches.
func matchLongestClassName(runes []rune, p int, names []string) (string, int) {
	for _, name := range names {
		nameRunes := []rune(name)
		if p+len(nameRunes) <= len(runes) && equalRunes(runes[p:p+len(nameRunes)], nameRunes) {
			return name, len(nameRunes)
		}
	}
	return "", 0
}

func (e *Engine) class(k tokenKind) [][]rune {
	if k == tokVowel {
		return e.vowels
	}
	return e.consonants
}

// matchPrefixClass returns the rune length of the longest class member that is a
// prefix of runes[p:], or -1 if none matches.
func matchPrefixClass(runes []rune, p int, set [][]rune) int {
	for _, m := range set {
		if p+len(m) <= len(runes) && equalRunes(runes[p:p+len(m)], m) {
			return len(m)
		}
	}
	return -1
}

// matchSuffixClass returns the rune length of the longest class member that is a
// suffix of runes[:end], or -1 if none matches.
func matchSuffixClass(runes []rune, end int, set [][]rune) int {
	for _, m := range set {
		if end-len(m) >= 0 && equalRunes(runes[end-len(m):end], m) {
			return len(m)
		}
	}
	return -1
}

func equalRunes(a, b []rune) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// rightMatch reports whether the right-context tokens match starting at p.
// Contexts are zero-width: we only verify, never alter the word.
func (e *Engine) rightMatch(runes []rune, p int, toks []token) bool {
	for _, t := range toks {
		switch t.kind {
		case tokBoundary:
			if p != len(runes) {
				return false
			}
		case tokVowel, tokConsonant:
			n := matchPrefixClass(runes, p, e.class(t.kind))
			if n < 0 {
				return false
			}
			p += n
		case tokClass:
			n := matchPrefixClass(runes, p, t.class)
			if n < 0 {
				return false
			}
			p += n
		case tokAny:
			if p >= len(runes) {
				return false
			}
			p++
		case tokLiteral:
			if p+len(t.lit) > len(runes) || !equalRunes(runes[p:p+len(t.lit)], t.lit) {
				return false
			}
			p += len(t.lit)
		}
	}
	return true
}

// leftMatch reports whether the left-context tokens match ending exactly at end.
// Tokens read left-to-right but anchor to the right edge, so we consume backward.
func (e *Engine) leftMatch(runes []rune, end int, toks []token) bool {
	for i := len(toks) - 1; i >= 0; i-- {
		t := toks[i]
		switch t.kind {
		case tokBoundary:
			if end != 0 {
				return false
			}
		case tokVowel, tokConsonant:
			n := matchSuffixClass(runes, end, e.class(t.kind))
			if n < 0 {
				return false
			}
			end -= n
		case tokClass:
			n := matchSuffixClass(runes, end, t.class)
			if n < 0 {
				return false
			}
			end -= n
		case tokAny:
			if end <= 0 {
				return false
			}
			end--
		case tokLiteral:
			if end-len(t.lit) < 0 || !equalRunes(runes[end-len(t.lit):end], t.lit) {
				return false
			}
			end -= len(t.lit)
		}
	}
	return true
}

// ApplyRule applies one rule to a word, left-to-right. Targets are consumed;
// environments are zero-width assertions against the original word.
//
// If r.Target names a registered user class, any one member of that class
// matches at the target position (longest member first), and the whole
// matched member is replaced by r.Replacement — the class acts as a
// whole-word target, not just an environment token. Otherwise r.Target is
// matched literally, exactly as before user classes existed.
func (e *Engine) ApplyRule(word string, r Rule) string {
	if !r.Enabled || r.Target == "" {
		return word
	}
	runes := []rune(word)
	left := tokenize(r.LeftEnv, e.userClassNames, e.userClasses)
	right := tokenize(r.RightEnv, e.userClassNames, e.userClasses)
	repl := []rune(r.Replacement)

	targetClass, isClassTarget := e.userClasses[r.Target]
	target := []rune(r.Target)

	var out []rune
	i := 0
	for i < len(runes) {
		var matchLen int
		if isClassTarget {
			matchLen = matchPrefixClass(runes, i, targetClass)
		} else if i+len(target) <= len(runes) && equalRunes(runes[i:i+len(target)], target) {
			matchLen = len(target)
		} else {
			matchLen = -1
		}

		if matchLen >= 0 &&
			e.leftMatch(runes, i, left) &&
			e.rightMatch(runes, i+matchLen, right) {
			out = append(out, repl...)
			i += matchLen
		} else {
			out = append(out, runes[i])
			i++
		}
	}
	return string(out)
}

// ApplyPipeline applies rules in order, recording which ones fired.
func (e *Engine) ApplyPipeline(word string, rules []Rule) Result {
	current := word
	applied := []string{}
	for _, r := range rules {
		if !r.Enabled {
			continue
		}
		next := e.ApplyRule(current, r)
		if next != current {
			applied = append(applied, r.Raw)
			current = next
		}
	}
	return Result{Original: word, Changed: current, RulesApplied: applied}
}

// BatchApply runs a pipeline over many words.
func (e *Engine) BatchApply(words []string, rules []Rule) []Result {
	out := make([]Result, len(words))
	for i, w := range words {
		out[i] = e.ApplyPipeline(w, rules)
	}
	return out
}
